File: android/vault-access/src/main/kotlin/org/secretary/browse/SettingsConversions.kt
package org.secretary.browse

/**
 * Milliseconds per minute — the minutes↔ms conversion base for the re-auth grace control.
 * ([MS_PER_DAY], the retention base, lives in `TrashFormatting`.)
 */
const val MS_PER_MINUTE: Long = 60_000L

// ---- Retention (days) ↔ ms — reuse msToDays (parity with desktop round(ms / MS_PER_DAY) + iOS) ----

/** Whole days in [ms], rounded to nearest, as an `Int` for the stepper (reuses [msToDays]). */
fun retentionDaysFromMs(ms: Long): Int = msToDays(ms).toInt()

/** Days → ms (parity with desktop `days * MS_PER_DAY`). A negative input clamps to 0 (the stepper
 *  never produces one; belt-and-suspenders for `Int`→`Long`). */
fun msFromRetentionDays(days: Int): Long = maxOf(0, days).toLong() * MS_PER_DAY

// ---- Re-auth grace (minutes) ↔ ms (parity with desktop round(ms / MS_PER_MINUTE) + iOS) ----

/** Whole minutes in [ms], rounded to nearest, as an `Int` for the stepper. */
fun graceMinutesFromMs(ms: Long): Int = ((ms + MS_PER_MINUTE / 2) / MS_PER_MINUTE).toInt()

/** Minutes → ms (parity with desktop `minutes * MS_PER_MINUTE`). Negative → 0. */
fun msFromGraceMinutes(minutes: Int): Long = maxOf(0, minutes).toLong() * MS_PER_MINUTE

// ---- Client-side clamps — validate against the projected FFI bounds (same constants the server enforces) ----

/** Clamp a retention-days input to the bounds' day range (inclusive). The bounds come from the port
 *  ([SettingsPort.settingsBounds]), so the UI validates against the same constants the server enforces. */
fun clampRetentionDays(days: Int, bounds: SettingsBounds): Int {
    val lo = retentionDaysFromMs(bounds.retentionMinMs)
    val hi = retentionDaysFromMs(bounds.retentionMaxMs)
    return minOf(maxOf(days, lo), hi)
}

/** Clamp a grace-minutes input to the bounds' minute range (inclusive). */
fun clampGraceMinutes(minutes: Int, bounds: SettingsBounds): Int {
    val lo = graceMinutesFromMs(bounds.reauthGraceMinMs)
    val hi = graceMinutesFromMs(bounds.reauthGraceMaxMs)
    return minOf(maxOf(minutes, lo), hi)
}

/** The banner shown after a successful settings save (mirror iOS `settingsSavedBanner`). */
fun settingsSavedBanner(): SettingsBanner = SettingsBanner("Settings saved")

File: ios/SecretaryVaultAccess/Sources/SecretaryVaultAccess/SettingsConversions.swift
import Foundation

/// Milliseconds per minute — the minutes↔ms conversion base for the re-auth
/// grace control. (`msPerDay`, the retention base, lives in `TrashFormatting`.)
public let msPerMinute: UInt64 = 60_000

// MARK: - Retention (days) ↔ ms

/// Whole days in `ms`, rounded to nearest, as an `Int` for the SwiftUI stepper.
/// Reuses `msToDays` (parity with desktop `Math.round(ms / MS_PER_DAY)`).
public func retentionDaysFromMs(_ ms: UInt64) -> Int { Int(msToDays(ms)) }

/// Days → ms (parity with desktop `days * MS_PER_DAY`). A negative input clamps
/// to 0 (the stepper never produces one; belt-and-suspenders for `Int`→`UInt64`).
public func msFromRetentionDays(_ days: Int) -> UInt64 { UInt64(max(0, days)) * msPerDay }

// MARK: - Re-auth grace (minutes) ↔ ms

/// Whole minutes in `ms`, rounded to nearest, as an `Int` for the stepper
/// (parity with desktop `Math.round(ms / MS_PER_MINUTE)`).
public func graceMinutesFromMs(_ ms: UInt64) -> Int { Int((ms + msPerMinute / 2) / msPerMinute) }

/// Minutes → ms (parity with desktop `minutes * MS_PER_MINUTE`). Negative → 0.
public func msFromGraceMinutes(_ minutes: Int) -> UInt64 { UInt64(max(0, minutes)) * msPerMinute }

// MARK: - Client-side clamps (validate against the projected FFI bounds)

/// Clamp a retention-days input to the bounds' day range (inclusive). The bounds
/// come from the port (`settingsBounds()`), so the UI validates against the same
/// constants the server enforces.
public func clampRetentionDays(_ days: Int, bounds: SettingsBounds) -> Int {
    let lo = retentionDaysFromMs(bounds.retentionMinMs)
    let hi = retentionDaysFromMs(bounds.retentionMaxMs)
    return min(max(days, lo), hi)
}

/// Clamp a grace-minutes input to the bounds' minute range (inclusive).
public func clampGraceMinutes(_ minutes: Int, bounds: SettingsBounds) -> Int {
    let lo = graceMinutesFromMs(bounds.reauthGraceMinMs)
    let hi = graceMinutesFromMs(bounds.reauthGraceMaxMs)
    return min(max(minutes, lo), hi)
}

// MARK: - Saved-confirmation banner

/// Inline saved-confirmation banner for the Settings screen (mirrors the Trash
/// `PurgeNotice` idiom). A settings save has no partial-failure state, so the
/// banner is always a benign confirmation; failures surface separately via the
/// view model's `error`.
public struct SettingsBanner: Equatable, Sendable {
    public let text: String
    public init(text: String) { self.text = text }
}

/// The banner shown after a successful settings save.
public func settingsSavedBanner() -> SettingsBanner { SettingsBanner(text: "Settings saved") }
